### ISSUE TITLE
Increased request latency in 0.106.1

### ISSUE THREAD

**Issue by arielvalentin** [open-telemetry/opentelemetry-collector#10858]
**Describe the bug**

_This issue was originally reported in CNCF slack by [Srinivas Bevara](https://cloud-native.slack.com/archives/C01N6P7KR6W/p1722488402533149) and confirmed by @codeboten who asked me to open this issue_

After Upgrade to v0.106.1 from v0.104.0 we have run into issues where we see an increase in request latency being reported by both the collector's internal telemetry and our ingress proxies. 

Memory remains unaffected however pprof is reporting garbage collection seems to have increased due internal telemetry tracing related allocations:

![Image](https://github.com/user-attachments/assets/cb91c276-348e-44ac-9526-82ebd8ed81f5)

![Image](https://github.com/user-attachments/assets/92038c0d-d8f4-4d06-aa45-4d61c36d8024)

![Image](https://github.com/user-attachments/assets/4a763cda-e23e-40e2-8861-5f51560f5f2c)

![Image](https://github.com/user-attachments/assets/033e652a-e428-4000-9995-d1651a1f617e)

This is unexpected since tracing telemetry is disabled in our deployments. 

**Steps to reproduce**

Run a load test comparing v0.104.0 and v0.106.1 

**What did you expect to see?**

- p99 Request latency and throughput should be comparable between both versions ~74ms
- Internal telemetry must not generate recording spans when it is disabled

**What did you see instead?**

- An increase in p99 latency to > 1s
- Profiling reveals increased objected allocations related to internal telemetry span data 
- Profiling reveals an increase in time spent in GC

**What version did you use?**

v0.106.1

**What config did you use?**

Our configuration has many transformations and connector pipelines. The bare minimum you may likely need is the one you use for benchmarking. 

**Environment**

Ubuntu 20.04.1

**Additional context**

Screen shots of increased p99 latency and CPU and steady memory 

![Image](https://github.com/user-attachments/assets/a3d7d33b-d29c-4732-bc0e-3a02e101e1cf)

![Image](https://github.com/user-attachments/assets/442fccb6-032e-477e-ad50-7d69c5ddfdcc)

![Image](https://github.com/user-attachments/assets/5b646a03-1b19-45c4-abb1-5d4542f03e09)

**Comment by codeboten** [comment]
After some discussion amongst maintainers, we want to provide a quick fix for this issue via the feature gate `service.noopTracerProvider`, and further investigate what's the cause of this problem before deciding if there should be a permanent configuration option to set a no-op tracer provider.

Returning a noop tracer provider if no tracing is configured in the `telemetry` settings would break the current behaviour, which the zpages extension relies on to provide telemetry about traces.

**Comment by arielvalentin** [comment]
Was there a recent change in zpages results in an increase in the number spans being allocated?

**Comment by bvsvas** [comment]
Thanks, @arielvalentin, for taking this up and raising a GitHub issue.

**Comment by arielvalentin** [comment]
@codeboten did this feature flag make it into v0.107.0?

**Comment by arielvalentin** [comment]
I recently upgraded to v0.107.0 from v0.104.0. During off peak times performance is about on-par but during our high peak times our p99 skyrockets and memory spikes forcing the memory limiter to trigger hard limits responding with 503s.  

I have enabled the feature gate to ensure we are using the noop-tracer but there still seems to be an increase in memory allocations and latency. 

Our heavy usage of OTTL has not changed much between versions and as far as I can tell that is what appears the most in the profiles along with a high number of context objects being allocated by the noop-tracer. 

[profiles-20240823.tar.gz](https://github.com/user-attachments/files/16737047/profiles-20240823.tar.gz)

**Comment by codeboten** [comment]
Re-opening to continue the investigation

**Comment by TylerHelmuth** [comment]
@arielvalentin Did the OTTL performance degrade, or are you saying it remained the majority?

**Comment by arielvalentin** [comment]
The profiles revealed that in 104 seemed to show OTTL execute was where it spent most of its time and allocations but our performance was pretty good ~15K spans per second per collector. p99 19ms.

Other than seeing the increased allocations to context objects related to nooptracer, the profiles between 104 and 107 look about the same to me so I don't have a reason to believe there is a regression in OTTL performance. 

I'll see if I can dig up my 104 profiles and attach them as well.

**Comment by pavolloffay** [comment]
From @arielvalentin 

> I have enabled the feature gate to ensure we are using the noop-tracer but there still seems to be an increase in memory allocations and latency.

> Other than seeing the increased allocations to context objects related to nooptracer, the profiles between 104 and 107 look about the same to me so I don't have a reason to believe there is a regression in OTTL performance.

So can be `service.noopTracerProvider` considered as a workaround?

**Comment by iblancasa** [comment]
> So can be `service.noopTracerProvider` considered as a workaround?

I would say "no" because
>  there still seems to be an increase in memory allocations and latency.

**Comment by codeboten** [comment]
Investigating this further, this might be related to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33508

**Comment by codeboten** [comment]
I just ran load tests for v0.107.0 and here's what i have when i enabled the noop tracer provider feature gate:

Before using `--feature-gates service.noopTracerProvider`

![Screenshot 2024-08-26 at 11 39 03 AM](https://github.com/user-attachments/assets/8c18b8bc-8879-4c3e-a34c-10e8a5ac74ca)

With `--feature-gates service.noopTracerProvider`

![Screenshot 2024-08-26 at 11 38 00 AM](https://github.com/user-attachments/assets/7122a793-ff9b-4370-ac58-79811a922d35)

This seems different than what you're seeing @arielvalentin, can you confirm that the feature gate is enabled?

**Comment by arielvalentin** [comment]
Yes. We currently run it with `--feature-gates service.noopTracerProvider`

You can see in the profiles I attached the noopTracer is used:

![Image](https://github.com/user-attachments/assets/47f4316e-7758-48c7-84d8-88c6e783ad01)

**Comment by arielvalentin** [comment]
Here are profiles for 104 [profiles-104.tar.gz](https://github.com/user-attachments/files/16753916/profiles-104.tar.gz)

**Comment by codeboten** [comment]
thanks @arielvalentin, I missed that when i looked your attached profiles a couple of days ago. good to know that using the no-op tracer provider at least reduces the overall memory consumption, even if its still higher than before the change to add tracing to ottl.

**Comment by arielvalentin** [comment]
No worries @codeboten 

Something I may have not mentioned already, but my colleague @antonio had observed the latency regression as early as 105.

We may downgrade to 105 and collect profiles there as well.

**Comment by antonio** [comment]
> Something I may have not mentioned already, but my colleague @antonio had observed the latency regression as early as 105.

That's correct. Notice how request duration went significantly up in the graph below

![Image](https://github.com/user-attachments/assets/444bb333-810a-47ee-85c7-1b982766a4f5)

> Investigating this further, this might be related to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33508

The time of when this PR was merged does indeed align with when we started seeing the issue.

**Comment by BinaryFissionGames** [comment]
Just adding what I've found here real quick;

I've done a quick benchmark to see the difference between an actual tracer impl, vs noop tracer, vs all tracing removed in ottl via the transform processor. You can see the benchmark [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/compare/b8423d696a4f7ae0276c268c021e5da46a09607c..observIQ:opentelemetry-collector-contrib:fix/tracing-mem-ottl#diff-21a3ef405eafe9c34655665f74a0d8c8b666b1cd2867f22da2c31b9c7ebb7d52R153). All it does is run the transform processor with a batch of 100 logs, with a simple set operation (which seems like a pretty normal workload).

These are the results:
```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor
cpu: Apple M3 Pro
                 │ no-tracing.txt │           noop-tracer.txt            │           test-tracer.txt            │
                 │     sec/op     │    sec/op     vs base                │    sec/op     vs base                │
SetLogRecords-12      50.29µ ± 1%   120.55µ ± 4%  +139.70% (p=0.002 n=6)   426.21µ ± 5%  +747.43% (p=0.002 n=6)

                 │ no-tracing.txt │            noop-tracer.txt            │            test-tracer.txt            │
                 │      B/op      │     B/op       vs base                │     B/op       vs base                │
SetLogRecords-12     96.21Ki ± 0%   227.45Ki ± 0%  +136.42% (p=0.002 n=6)   758.38Ki ± 1%  +688.28% (p=0.002 n=6)

                 │ no-tracing.txt │           noop-tracer.txt           │           test-tracer.txt           │
                 │   allocs/op    │  allocs/op   vs base                │  allocs/op   vs base                │
SetLogRecords-12      1.402k ± 0%   3.802k ± 0%  +171.18% (p=0.002 n=6)   9.802k ± 0%  +599.14% (p=0.002 n=6)
```

You can see that even the noop tracer increases these benchmarks by quite a bit. And it adds up with what I'm seeing in both the [issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34890) I posted and here, which is that the overhead of tracing each individual evaluation of the StatementSequence seems like overkill. 

Of course, if you do more processing in a single ottl statement, you'd probably get numbers that look a little less striking, so take the exact numbers here with a grain of salt. I don't think the scenario is all too unlikely, though.

It feels like there should not be tracing in StatementSequence at all, but maybe a few metric counters instead. The overhead of the tracing just doesn't seem worth it.

**Comment by arielvalentin** [comment]
I have got some more for you here [profiles-20240828.tar.gz](https://github.com/user-attachments/files/16785499/profiles-20240828.tar.gz)

cc: @songy23

**Comment by TylerHelmuth** [comment]
I am going to submit a PR to revert the ottl tracing stuff.

**Comment by iblancasa** [comment]
> I am going to submit a PR to revert the ottl tracing stuff.

How about placing it behind a feature gate?

**Comment by arielvalentin** [comment]
Thanks again for investigating. I am on the path to downgrade to 0.104.0 and will be eagerly awaiting the next release. 

Once it's out, I will deploy to a subset of our canary instances and report back profiles and latency info.

**Comment by arielvalentin** [comment]
@bvsvas have you all had a chance to generate profiles? 

IIRC you all were not using OTTL so the latency you all reported would not be related to the transformers right? 

This is the snippet you posted in slack:

```yaml
service:
  extensions: [apmauth, pprof]
  pipelines:
    traces:
      receivers: [otlp/withauth]
      processors: [probabilistic_sampler]
      exporters: [customexporter]
    metrics:
      receivers: [otlp/withauth, prometheus]
      processors: [cumulativetodelta, resource]
      exporters: [customexporter]
```

**Comment by TylerHelmuth** [comment]
An argument for moving `component.Telemetry` and `component.BuildInfo` somewhere else is that they have no dependency on `component` so moving them to their own package could save us from a circular dependency later.

**Comment by arielvalentin** [comment]
👋🏼 Is there a release schedule/project board handy I can subscribe to?

**Comment by codeboten** [comment]
The release schedule is documented here: https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/release.md#release-schedule

**Comment by deepakd10** [comment]
I also faced this same bug, Another work around. Set `gomemlimit`, So that GC is kicked in to free up the mem.

https://weaviate.io/blog/gomemlimit-a-game-changer-for-high-memory-applications#adding-gomemlimit-to-weaviate--before--after
